File: aws_batch_squared_py/vpc_stack.py
"""VPC nested stack — creates a new VPC or imports an existing one.

Ported from ``lib/vpc-stack.ts``.
"""

from typing import List, Optional

import aws_cdk as cdk
from aws_cdk import aws_ec2 as ec2
from constructs import Construct


class VpcStack(cdk.NestedStack):
    vpc: ec2.IVpc
    subnets: List[ec2.ISubnet]

    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        *,
        namespace: str,
        create_vpc: bool,
        existing_vpc_id: Optional[str] = None,
        existing_subnet_ids: Optional[List[str]] = None,
        **kwargs,
    ) -> None:
        super().__init__(scope, construct_id, **kwargs)

        if create_vpc:
            # Create new VPC
            vpc = ec2.Vpc(
                self,
                "VPC",
                ip_addresses=ec2.IpAddresses.cidr("10.0.0.0/16"),
                max_azs=2,
                nat_gateways=2,
                subnet_configuration=[
                    ec2.SubnetConfiguration(
                        cidr_mask=24,
                        name=f"{namespace}-public",
                        subnet_type=ec2.SubnetType.PUBLIC,
                    ),
                    ec2.SubnetConfiguration(
                        cidr_mask=24,
                        name=f"{namespace}-private",
                        subnet_type=ec2.SubnetType.PRIVATE_WITH_EGRESS,
                    ),
                ],
                enable_dns_hostnames=True,
                enable_dns_support=True,
            )

            cdk.Tags.of(vpc).add("Name", f"{namespace}-vpc")

            self.vpc = vpc
            self.subnets = vpc.private_subnets
        else:
            # Use existing VPC
            if not existing_vpc_id or not existing_subnet_ids:
                raise ValueError(
                    "existingVpcId and existingSubnetIds are required when createVpc is false"
                )

            self.vpc = ec2.Vpc.from_lookup(self, "ExistingVpc", vpc_id=existing_vpc_id)

            self.subnets = [
                ec2.Subnet.from_subnet_id(self, f"ExistingSubnet{index}", subnet_id)
                for index, subnet_id in enumerate(existing_subnet_ids)
            ]

        # Outputs
        cdk.CfnOutput(self, "VpcId", value=self.vpc.vpc_id)
        cdk.CfnOutput(
            self,
            "SubnetIds",
            value=",".join(s.subnet_id for s in self.subnets),
        )

File: playbook/ways-to-run/patterns/batch-squared/infrastructure/typescript/lib/vpc-stack.ts
import * as cdk from "aws-cdk-lib";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import type { Construct } from "constructs";

export interface VpcStackProps extends cdk.NestedStackProps {
  namespace: string;
  createVpc: boolean;
  existingVpcId?: string;
  existingSubnetIds?: string[];
}

export class VpcStack extends cdk.NestedStack {
  public readonly vpc: ec2.IVpc;
  public readonly subnets: ec2.ISubnet[];

  constructor(scope: Construct, id: string, props: VpcStackProps) {
    super(scope, id, props);

    if (props.createVpc) {
      // Create new VPC
      const vpc = new ec2.Vpc(this, "VPC", {
        ipAddresses: ec2.IpAddresses.cidr("10.0.0.0/16"),
        maxAzs: 2,
        natGateways: 2,
        subnetConfiguration: [
          {
            cidrMask: 24,
            name: `${props.namespace}-public`,
            subnetType: ec2.SubnetType.PUBLIC,
          },
          {
            cidrMask: 24,
            name: `${props.namespace}-private`,
            subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,
          },
        ],
        enableDnsHostnames: true,
        enableDnsSupport: true,
      });

      cdk.Tags.of(vpc).add("Name", `${props.namespace}-vpc`);

      // S3 Gateway Endpoint — routes S3 traffic directly, bypassing NAT (free, no data charges)
      vpc.addGatewayEndpoint("S3Endpoint", {
        service: ec2.GatewayVpcEndpointAwsService.S3,
      });

      this.vpc = vpc;
      this.subnets = vpc.privateSubnets;
    } else {
      // Use existing VPC
      if (!props.existingVpcId || !props.existingSubnetIds) {
        throw new Error(
          "existingVpcId and existingSubnetIds are required when createVpc is false",
        );
      }

      this.vpc = ec2.Vpc.fromLookup(this, "ExistingVpc", {
        vpcId: props.existingVpcId,
      });

      this.subnets = props.existingSubnetIds.map((subnetId, index) =>
        ec2.Subnet.fromSubnetId(this, `ExistingSubnet${index}`, subnetId),
      );
    }

    // Outputs
    new cdk.CfnOutput(this, "VpcId", {
      value: this.vpc.vpcId,
    });

    new cdk.CfnOutput(this, "SubnetIds", {
      value: this.subnets.map((s) => s.subnetId).join(","),
    });
  }
}
